% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/archipelago.R
\name{archipelago_plot}
\alias{archipelago_plot}
\title{Archipelago Plot}
\arguments{
\item{df1}{A dataframe of variant-set data.}

\item{df2}{A dataframe of variant data.}

\item{plot_title}{Title for the plot.}

\item{add_title}{Logical; add title if TRUE.}

\item{plot_subtitle}{Subtitle for the plot.}

\item{add_subtitle}{Logical; add subtitle if TRUE.}

\item{chr_ticks}{Logical; show chromosome ticks if TRUE.}

\item{show_legend}{Logical; display the legend if TRUE.}

\item{color_theme}{Name of the colour theme.}

\item{custom_colors}{Vector of custom colours.}

\item{color_labels}{Labels for the colour groups.}

\item{crit_val}{Critical p-value threshold.}

\item{point_size}{Size of the points.}

\item{point_size_large}{Size of the large points of interest such as VSAT.}

\item{fig_width}{Width of the archipelago plot.}

\item{fig_height}{Height of the archipelago plot.}

\item{raw_fig_width}{Width of the raw plot.}

\item{raw_fig_height}{Height of the raw plot.}

\item{output_path}{File path for the plot.}

\item{output_raw}{File path for the raw plot.}

\item{file_type}{Use .png, .jpg, or .pdf. Defaults to png. pdf is slow and large for many SNPs.}

\item{alpha_point}{Use the alpha_point value to set point transparency.}

\item{alpha_seg}{Use the alpha_segvalue to set line segment transparency.}

\item{better_space}{Use better_space = TRUE to make sure that x-axis chr do not squash.}

\item{legend_position}{Default right, allows bottom top left right.}
}
\value{
A ggplot object.
}
\description{
Plot p-values from variant and variant-set tests.
}
\examples{
\dontrun{
  # Load example data for df1 (VSAT) and df2 (individual variant)
  data("vsat_pval")         # same structure as df1
  data("variant_pval")      # same structure as df2

  # Basic usage with defaults
  archipelago_plot(
    df1 = vsat_pval,
    df2 = variant_pval
  )

  # Specify a built-in colour theme
  archipelago_plot(
    df1 = vsat_pval,
    df2 = variant_pval,
    color_theme = "alice"
  )

  # More customised usage
  output_path <- "./archipelago_plot_custom_color.pdf"
  output_raw <- "./vsat_raw_plot.pdf"
  color_labels <- c("Label_1", "Label_2", "Label_3", "Label_4")
  custom_colors <- c("#9abfd8", "#cac1f3", "#371c4b", "#2a5b7f")

  archipelago_plot(
    df1 = vsat_pval,
    df2 = variant_pval,
    add_title = TRUE,
    plot_title = "My Archipelago Plot",
    add_subtitle = TRUE,
    plot_subtitle = "VSAT vs Single Variant",
    show_legend = TRUE,
    chr_ticks = FALSE,
    point_size = 0.5,
    color_theme = NULL,         # ignore built-in theme
    custom_colors = custom_colors,
    color_labels = color_labels,
    crit_val_VSAT = 0.05 / 300, # for highlighting VSAT p-values
    crit_val_single_variant = 5e-8, # typical single-variant threshold
    output_path = output_path, 
    output_raw = output_raw,   
    file_type = "pdf"          # save as PDF instead of PNG
  )
}
}
