<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Method for Illustration of Variant Set Association Test Statistics â€¢ archipelago</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Method for Illustration of Variant Set Association Test Statistics">
<meta property="og:description" content="Provides a graphical method for interpreting both VSAT p-values and individual variant contributions. The Archipelago method assigns a meaningful genomic coordinate to the VSAT p- value, enabling its simultaneous visualization alongside individual variant p-values. This results in an intuitive and rich illustration akin to an archipelago of clustered islands, enhancing the understanding of both collective and individual impacts of variants. The Archipelago plot is applicable in any genetic association study that uses variant collapse to evaluate both individual variants and variant sets, and its customisability facilitates clear communication of complex genetic data. By integrating two dimensions of genetic data into a single visualization, VSAT results can be easily read and aid in identification of potential causal variants in variant sets such as protein pathways.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">archipelago</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="contents col-md-9">

<div class="section level1">
<div class="page-header"><h1 id="archipelago-plot">Archipelago plot<a class="anchor" aria-label="anchor" href="#archipelago-plot"></a>
</h1></div>
<div class="section level2">
<h2 id="summary-and-illustration-of-variant-set-association-test-statistics">Summary and illustration of variant set association test statistics<a class="anchor" aria-label="anchor" href="#summary-and-illustration-of-variant-set-association-test-statistics"></a>
</h2>
<p>Variant set association tests (VSAT), particularly those incorporating minor allele frequency variants, have become invaluable in genetic association studies by allowing robust statistical analysis with variant collapse. Unlike single variant tests, VSAT statistics cannot be assigned to a genomic coordinate for visual interpretation by default. To address these challenges, we introduce the Archipelago plot, a graphical method for interpreting both VSAT p-values and individual variant contributions. The Archipelago method assigns a meaningful genomic coordinate to the VSAT p-value, enabling its simultaneous visualization alongside individual variant p-values. This results in an intuitive and rich illustration akin to an archipelago of clustered islands, enhancing the understanding of both collective and individual impacts of variants. The Archipelago plot is applicable in any genetic association study that uses variant collapse to evaluate both individual variants and variant sets, and its customizability facilitates clear communication of complex genetic data. By integrating two dimensions of genetic data into a single visualization, VSAT results can be easily read and aid in identification of potential causal variants in variant sets such as protein pathways.</p>
<div class="float">
<img src="./images/DALLE_Archipelago.jpg" alt="Archipelago cover image"><div class="figcaption">Archipelago cover image</div>
</div>
<div class="float">
<img src="./images/Archipelago_plot_5kvar_250set_legend_v1.jpg" alt="Default plot image"><div class="figcaption">Default plot image</div>
</div>
</div>
<div class="section level2">
<h2 id="how-to">How to<a class="anchor" aria-label="anchor" href="#how-to"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Install the package</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"~/archipelago_0.0.0.9000.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">archipelago</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load test data for df1 and df2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"vsat_pval"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"variant_pval"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Import user data</span></span>
<span><span class="co"># df1 &lt;- read.csv(file="../data/vsat_pval.txt")</span></span>
<span><span class="co"># df2 &lt;- read.csv(file="../data/variant_pval.txt")</span></span>
<span></span>
<span><span class="co"># Use default settings</span></span>
<span><span class="fu"><a href="reference/archipelago_plot.html">archipelago_plot</a></span><span class="op">(</span><span class="va">df1</span>, <span class="va">df2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 16 color themes:</span></span>
<span><span class="co"># 'retro', 'metro', 'summer', 'messenger', 'sunset', 'alice', 'buckley', 'romance', 'meme', 'saiko', 'pagliacci', 'ambush', 'sunra', 'caliber', 'yawn', 'lawless'</span></span>
<span></span>
<span><span class="fu"><a href="reference/archipelago_plot.html">archipelago_plot</a></span><span class="op">(</span><span class="va">df1</span>,</span>
<span>                 <span class="va">df2</span>,</span>
<span>                 color_theme <span class="op">=</span> <span class="st">'alice'</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Custom everything</span></span>
<span><span class="va">output_path</span> <span class="op">=</span> <span class="st">"./archipelago_plot_custom_color.pdf"</span></span>
<span><span class="va">output_raw</span> <span class="op">=</span> <span class="st">"./vsat_raw_plot.pdf"</span></span>
<span><span class="va">color_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Label_1"</span>, <span class="st">"Label_2"</span> ,<span class="st">"Label_3"</span>,<span class="st">"Label_4"</span><span class="op">)</span></span>
<span><span class="va">custom_colors</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#9abfd8"</span>, <span class="st">"#cac1f3"</span>, <span class="st">"#371c4b"</span>, <span class="st">"#2a5b7f"</span><span class="op">)</span> <span class="co"># buckley theme colors</span></span>
<span><span class="va">plot_title</span> <span class="op">&lt;-</span> <span class="st">"Title"</span></span>
<span><span class="va">plot_subtitle</span> <span class="op">&lt;-</span> <span class="st">"Subtitle"</span></span>
<span><span class="va">crit_val</span> <span class="op">=</span> <span class="fl">.05</span><span class="op">/</span><span class="fl">300</span> <span class="co"># P-value threshold line</span></span>
<span><span class="va">point_size</span> <span class="op">=</span> <span class="fl">.5</span> <span class="co"># geom_point size</span></span>
<span><span class="va">output_path</span> <span class="op">=</span> <span class="st">"../output/archipelago_plot_custom_everything.pdf"</span></span>
<span><span class="fu"><a href="reference/archipelago_plot.html">archipelago_plot</a></span><span class="op">(</span>df1 <span class="op">=</span> <span class="va">df1</span>,</span>
<span>                 d <span class="op">=</span> <span class="va">df2</span>,</span>
<span>                 add_title <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 plot_title <span class="op">=</span> <span class="va">plot_title</span>,</span>
<span>                 add_subtitle <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 plot_subtitle <span class="op">=</span> <span class="va">plot_subtitle</span>,</span>
<span>                 show_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 crit_val <span class="op">=</span> <span class="va">crit_val</span>,</span>
<span>                 point_size <span class="op">=</span> <span class="va">point_size</span>,</span>
<span>                 <span class="va">chr_ticks</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span>, <span class="co"># TRUE / FALSE</span></span>
<span>                 <span class="va">show_legend</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span>, <span class="co"># TRUE / FALSE</span></span>
<span>                 custom_colors <span class="op">=</span> <span class="va">custom_colors</span>,</span>
<span>                 color_labels <span class="op">=</span> <span class="va">color_labels</span>,</span>
<span>                 output_path <span class="op">=</span> <span class="va">output_path</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level1">
<h1 id="dev-notes">Dev notes<a class="anchor" aria-label="anchor" href="#dev-notes"></a>
</h1>
<div class="section level2">
<h2 id="set-up">Set up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h2>
<pre><code>Rscript -e "usethis::create_package(getwd())"</code></pre>
</div>
<div class="section level2">
<h2 id="docs-with">Docs with<a class="anchor" aria-label="anchor" href="#docs-with"></a>
</h2>
<pre><code>Rscript -e "devtools::document()"</code></pre>
</div>
<div class="section level2">
<h2 id="build-and-install">Build and install<a class="anchor" aria-label="anchor" href="#build-and-install"></a>
</h2>
<pre><code>Rscript -e "devtools::build()"
Rscript -e "devtools::install()"</code></pre>
</div>
<div class="section level2">
<h2 id="install-from-zip-in-r">Install from zip in R<a class="anchor" aria-label="anchor" href="#install-from-zip-in-r"></a>
</h2>
<pre><code><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"/path/to/your/package/MyPackage.tar.gz"</span>, repos <span class="op">=</span> <span class="cn">NULL</span>, type <span class="op">=</span> <span class="st">"source"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="example-of-set-id">Example of set ID<a class="anchor" aria-label="anchor" href="#example-of-set-id"></a>
</h2>
<p>An example of the protein pathway set IDs can be found at <a href="https://github.com/DylanLawless/ProteoMCLustR/tree/main/data/ppi_examples" class="external-link uri">https://github.com/DylanLawless/ProteoMCLustR/tree/main/data/ppi_examples</a> where we have a numeric pathway ID for each ensemble gene name, using MCL clustering on STRINGdb.</p>
<div class="float">
<img src="./images/archipelago_logo_v2.jpg" alt="Logo image"><div class="figcaption">Logo image</div>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing archipelago</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Dylan Lawless <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/https://orcid.org/0000-0001-8496-3725" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Dylan Lawless.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

      </footer>
</div>






  </body>
</html>
